tests/rust-plugins/swc_internal_plugin/src/lib.rs

```diff
@@ -36,10 +36,11 @@ impl VisitMut for ConsoleOutputReplacer {
 ///     i32 /* 0 means success */
 ///
 /// if plugin need to handle low-level ptr directly. However, there are important steps
-/// manually need to be performed like transformed results back to host. Refer swc_plugin_macro
+/// manually need to be performed like sending transformed results back to host. Refer swc_plugin_macro
 /// how does it work internally.
 #[plugin_module]
-pub fn process(program: Program, _plugin_config: String) -> Result<Program, PluginError> {
-    let transformed_program = program.fold_with(&mut as_folder(ConsoleOutputReplacer));
-    Ok(transformed_program)
+pub fn process(program: Program, _plugin_config: String, ctx: &HostContext) -> Program {
+    let test_diag = Diagnostic::new(Level::Error, "test_diag_error_from_plugin");
```

Comment:
What happens if a user use

```rust

HANDLER.with(|handler| {
     handler.struct_span_err(sp, "test").emit();
})
```

?

I think we should call user-defined transform function within a closure with `HANDLER` configured.

Reply:
I think it will be easy since the only required tasks are creating a `Handler` implementation which calls `HostContext.emit` on `Handler::emit`, and configuring `HANDLER` using `HANDLER.set`

Reply:
I'm not sure if I understood correctly, but `HANDLER` supposed to be unique in host's context, isn't it? Plugin does not have way to access those unique context inside of wasm.

Reply:
No, it's a scoped thread-local variable.
`HANDLER` is configured only in closures provided to `HANDLER.set`.

Reply:
The macro for plugin should configure one, and call the transform function in the closure provided to `HANDLER.set`

Reply:
Ok, I think I'm quite confused. As you said, HANDLER is scoped_thread_local and since plugin is wasm, it is not possible to compile plugin with wasm-* target if we import HANDLER, since wasm currently doesn't have thread support as well as scoped-tls is not able to target wasm-*.

Maybe I'm missing some obvious? 🤔

Reply:
hmm, I recall first time I tried to import `HANDLER` there was a build failure but somehow attempting once again it seems at least build. Not sure what's changed.

Back to the original question, would you mind elaborating thoughts around importing whole handler in the plugin side? I think I'm not getting full pictures correctly in here.

Reply:
I experimented, but seems like thread-local requires `target-feature=+atomic`, `target-feature=+mutable-globals`, `target-feature=+bulk-memory` and something more.
So... Using `scoped-tls` with wasm will be quite tricky at the moment.

I stopped the investigation, but I think we need a way to report errors/create private identifiers from the plugin, while allowing testing them.
User plugin will depend on `swc_common/plugin-mode`, but while testing we may need to use normal `swc_common`, not one for `plugin-mode`.


For this PR, I think this is the best we can do at the moment. Thank you!